### PR TITLE
Initial integration of shared cluster chart

### PR DESCRIPTION
**Pull request by glitchcrab**
Towards https://github.com/giantswarm/roadmap/issues/3124

I templated the chart out from main and then from this branch and compared the two directory structures with `diff`:

```
 ➜  diff -ur /tmp/vsphere/main/cluster-vsphere/templates/ /tmp/vsphere/feature/cluster-vsphere/templates/
 ➜
```

I then removed `.global.nodePools` from the values files on my branch and templated the branch out again to test usage of the defaults added in this PR (see notes below code block). Note that this doesn't include the `Cluster` resource - that is further down:

```
 ➜  diff -ur /tmp/vsphere/pre/cluster-vsphere/templates/ /tmp/vsphere/defaults/cluster-vsphere/templates/
diff -ur /tmp/vsphere/pre/cluster-vsphere/templates/kubeadmconfigtemplate.yaml /tmp/vsphere/defaults/cluster-vsphere/templates/kubeadmconfigtemplate.yaml
--- /tmp/vsphere/pre/cluster-vsphere/templates/kubeadmconfigtemplate.yaml	2024-07-19 12:11:29.769101218 +0100
+++ /tmp/vsphere/defaults/cluster-vsphere/templates/kubeadmconfigtemplate.yaml	2024-07-19 12:15:31.997789637 +0100
@@ -3,7 +3,7 @@
 apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
 kind: KubeadmConfigTemplate
 metadata:
-  name: test-worker-c275c4fe
+  name: test-def00-fa98c372
   namespace: default
   labels:
     app: "cluster-vsphere"
@@ -119,7 +119,7 @@
             eviction-soft-grace-period: "memory.available=5s"
             anonymous-auth: "true"

-            node-labels: "giantswarm.io/node-pool=worker"
+            node-labels: "giantswarm.io/node-pool=def00"
       files:
         - path: /etc/ssh/trusted-user-ca-keys.pem
           permissions: "0600"
diff -ur /tmp/vsphere/pre/cluster-vsphere/templates/machinedeployment.yaml /tmp/vsphere/defaults/cluster-vsphere/templates/machinedeployment.yaml
--- /tmp/vsphere/pre/cluster-vsphere/templates/machinedeployment.yaml	2024-07-19 12:11:29.769101218 +0100
+++ /tmp/vsphere/defaults/cluster-vsphere/templates/machinedeployment.yaml	2024-07-19 12:15:31.997789637 +0100
@@ -3,7 +3,7 @@
 apiVersion: cluster.x-k8s.io/v1beta1
 kind: MachineDeployment
 metadata:
-  name: test-worker
+  name: test-def00
   namespace: default
   labels:
     app: "cluster-vsphere"
@@ -35,12 +35,12 @@
         configRef:
           apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
           kind: KubeadmConfigTemplate
-          name: test-worker-c275c4fe
+          name: test-def00-fa98c372
           namespace: default
       clusterName: test
       infrastructureRef:
         apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
         kind: VSphereMachineTemplate
-        name: test-worker-81f02996
+        name: test-def00-1f2b4d56
         namespace: default
       version: v1.27.14
diff -ur /tmp/vsphere/pre/cluster-vsphere/templates/vspheremachinetemplate.yaml /tmp/vsphere/defaults/cluster-vsphere/templates/vspheremachinetemplate.yaml
--- /tmp/vsphere/pre/cluster-vsphere/templates/vspheremachinetemplate.yaml	2024-07-19 12:11:29.769101218 +0100
+++ /tmp/vsphere/defaults/cluster-vsphere/templates/vspheremachinetemplate.yaml	2024-07-19 12:15:31.997789637 +0100
@@ -36,7 +36,7 @@
 apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
 kind: VSphereMachineTemplate
 metadata:
-  name: test-worker-81f02996
+  name: test-def00-1f2b4d56
   namespace: default
   labels:
     app: "cluster-vsphere"
@@ -55,12 +55,8 @@
       server: https://foo.example.com
       thumbprint: F7:CF:F9:E5:99:39:FF:C1:D7:14:F1:3F:8A:42:21:95:3B:A1:6E:16
       cloneMode: linkedClone
-      diskGiB: 25
-      memoryMiB: 8192
-      network:
-        devices:
-        - dhcp4: true
-          networkName: grasshopper-capv
-      numCPUs: 2
-      resourcePool: grasshopper
+      memoryMiB: 16896
+      network: {}
+      numCPUs: 6
+      resourcePool: '*/Resources'
       template: flatcar-stable-3815.2.2-kube-v1.27.14-gs
```

Notes on the diff:

- the nodepool names change from `worker` to `def00` to align with other provider nodePool naming.
- the worker vspheremachinetemplate spec changes a little because the defaults are different to the values which are used in `main` at the moment. The templated resource is still a valid CR.
- changing the values in `cluster.providerIntegration.workers.defaultNodePools.def00` results in an updated `mtRevisionHash` so the hashing function still works correctly following the removal of `node`Classes` (not shown in the diff but tested locally).

## Cluster

Cluster resource rendered from `main`:

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: test
  namespace: default
  annotations:
    cluster.giantswarm.io/description: "test cluster"
  labels:
    cluster-apps-operator.giantswarm.io/watching: ""
    giantswarm.io/service-priority: "highest"
    app: "cluster-vsphere"
    app.kubernetes.io/managed-by: "Helm"
    cluster.x-k8s.io/cluster-name: "test"
    giantswarm.io/cluster: "test"
    giantswarm.io/organization: "giantswarm"
    application.giantswarm.io/team: "rocket"
    app.kubernetes.io/version: "0.56.1"
    helm.sh/chart: "cluster-vsphere-0.56.1"
    policy.giantswarm.io/psp-status: disabled
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
        - 10.244.0.0/16
    services:
      cidrBlocks:
        - 172.31.0.0/16
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: KubeadmControlPlane
    name: test
    namespace: default
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: VSphereCluster
    name: test
    namespace: default
````

Cluster resource rendered from this branch:

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  annotations:
    cluster.giantswarm.io/description: "test cluster"
  labels:
    cluster-apps-operator.giantswarm.io/watching: ""
    # deprecated: "app: cluster-vsphere" label is deprecated and it will be removed after upgrading
    # to Kubernetes 1.25. We still need it here because existing ClusterResourceSet selectors
    # need this label on the Cluster resource.
    app: "cluster-vsphere"
    app.kubernetes.io/name: "cluster"
    app.kubernetes.io/version: "0.35.0"
    app.kubernetes.io/part-of: "cluster-vsphere"
    app.kubernetes.io/instance: "test"
    app.kubernetes.io/managed-by: "Helm"
    helm.sh/chart: "cluster-0.35.0"
    application.giantswarm.io/team: "turtles"
    giantswarm.io/cluster: "test"
    giantswarm.io/organization: "giantswarm"
    giantswarm.io/service-priority: highest
    cluster.x-k8s.io/cluster-name: "test"
    cluster.x-k8s.io/watch-filter: capi
    policy.giantswarm.io/psp-status: disabled
  name: test
  namespace: org-t-eu1qa2b7j76eccn72v
spec:
  clusterNetwork:
    apiServerPort: 6443
    services:
      cidrBlocks:
        - 172.31.0.0/16
    pods:
      cidrBlocks:
        - 10.244.0.0/16
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: KubeadmControlPlane
    name: test
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: VSphereCluster
    name: test
```

## E2E tests

- `standard`: https://tekton.ci.giantswarm.io/#/namespaces/tekton-pipelines/pipelineruns/pr-cluster-vsphere-238-cluster-test-suites6g7fg?pipelineTask=run-tests&taskRunName=pr-cluster-vsphere-238-cluster-test-suites6g7fg-run-tests-1&step=run-tests
- `upgrade`: https://github.com/giantswarm/cluster-vsphere/pull/238#issuecomment-2243259840
- `on-capv`: https://github.com/giantswarm/cluster-vsphere/pull/238#issuecomment-2242990116

### CODE REVIEW

File: helm/cluster-vsphere/templates/_helpers.tpl [hunk 1]
```diff
@@ -52,7 +52,7 @@ create VSphereMachineTemplates.
 {{ define "createMapOfClusterNodeSpecs" }}
 {{- $nodeMap := dict -}}
 {{- $_ := set $nodeMap "control-plane" .Values.global.controlPlane.machineTemplate -}}
-{{- range $index, $pool := .Values.global.nodePools -}}
+{{- range $index, $pool := .Values.global.nodePools | default .Values.cluster.providerIntegration.workers.defaultNodePools -}}
```
Comment (glitchcrab) [inline helm/cluster-vsphere/templates/_helpers.tpl:55]: use the default worker nodepool configuration if none provided in values

File: helm/cluster-vsphere/templates/machinedeployment.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-{{- range $name, $value := include "createMapOfWorkerPoolSpecs" . | fromYaml }}
+{{- range $nodePoolName, $value := include "createMapOfWorkerPoolSpecs" . | fromYaml }}
```
Comment (glitchcrab) [inline helm/cluster-vsphere/templates/machinedeployment.yaml:1]: just aligning variable naming with other providers

File: helm/cluster-vsphere/values.schema.json [hunk 3]
```diff
@@ -116,12 +116,67 @@
...
+                        "infrastructureCluster": {
+                            "group": "infrastructure.cluster.x-k8s.io",
+                            "kind": "VSphereCluster",
+                            "version": "v1beta1"
+                        },
+                        "machineHealthCheckResourceEnabled": false,
+                        "machinePoolResourcesEnabled": false,
+                        "networkPoliciesHelmReleaseResourceEnabled": false,
+                        "nodePoolKind": "MachineDeployment",
+                        "verticalPodAutoscalerCrdHelmReleaseResourceEnabled": false
```
Comment (glitchcrab) [inline helm/cluster-vsphere/values.schema.json:153]: disable all resources from rendering apart from the `Cluster`

File: helm/cluster-vsphere/values.schema.json [hunk 4]
```diff
@@ -116,12 +116,67 @@
...
+                    "workers": {
+                        "defaultNodePools": {
+                            "def00": {
+                                "cloneMode": "linkedClone",
+                                "memoryMiB": 16896,
+                                "network": {},
+                                "numCPUs": 6,
+                                "replicas": 2,
+                                "resourcePool": "*/Resources",
+                                "template": "flatcar-stable-3815.2.2-kube-v1.27.14-gs"
```
Comment (glitchcrab) [inline helm/cluster-vsphere/values.schema.json:164]: default nodepool configuration if none is provided

**Comment by glitchcrab** [comment]
the `cmp-helm-rendering` check will fail because the CI values files have been updated with the `name` value (which isn't allowed in the `main` branch values schema.

**Comment by glitchcrab** [comment]
nope, forgot to update the e2e values. all tests will fail until the linked PR :point_up: is merged

**Comment by gawertm** [comment]
nice

**Review by vxav — APPROVED** [review]
If tested looks good to me. I can't see anything wrong with it. Good job 💪

**Comment by glitchcrab** [comment]
and of course the upgrade test will fail to deploy a cluster because of the added `name` value:

```
status:
  appVersion: ""
  release:
    lastDeployed: null
    reason: |
      values don't meet the specifications of the schema(s) in the following chart(s):
      cluster-vsphere:
      - global.metadata: Additional property name is not allowed
    status: values-schema-violation
```

**Comment by erkanerol** [comment]
@glitchcrab I checked the deletion issue. There was a problem with the cleaner operator. Restarting the pod solved the issue. I created a ticket for that to our board.

**Pull request merged**